[Cocoa] Inserting empty text using system text input interfaces does not delete the current selection

pxlcoder opened the pull request:
#### 75df801c372223a2dd89cabc04f4b7dde73d1cbd
<pre>
[Cocoa] Inserting empty text using system text input interfaces does not delete the current selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=263764">https://bugs.webkit.org/show_bug.cgi?id=263764</a>
rdar://106570343

Reviewed by Wenson Hsieh.

Passing in an empty string to `-[NSResponder insertText:]` and
`-[UITextInput insertText:]` while there is a ranged selection should
perform backwards deletion, matching other system text engines.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::insertTextAsync):

`Editor::insertTextWithoutSendingTextEvent` early returns when empty text is
passed in. In order to avoid affecting other codepaths, implement the logic in
`WebPage::insertTextAsync`, which a common entry point for system text input
insertion across Cocoa platforms.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewTextInput.mm:
(insertText):
(TEST):

Canonical link: <a href="https://commits.webkit.org/269858@main">https://commits.webkit.org/269858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2498dfd3f0a2d67ec4706f739e965f10ee177892

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21941 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22469 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1446 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20569 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26535 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1219 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21480 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27735 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21692 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21759 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25516 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1176 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18870 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1178 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5697 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1494 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6541,9 +6541,13 @@ void WebPage::insertTextAsync(const String& text, const EditingRange& replacemen
-        // An insertText: might be handled by other responders in the chain if we don't handle it.
-        // One example is space bar that results in scrolling down the page.
-        frame->editor().insertText(text, nullptr, replacesText ? TextEventInputAutocompletion : TextEventInputKeyboard);
+        if (text.isEmpty() && frame->selection().isRange())
```
Should this logic also be implemented in `insertDictatedTextAsync`? Though this may not even be possible to do while dictating

pxlcoder replied:
I don't think that's necessary, since `insertDictatedTextAsync` is only called if there are alternatives, and you can't have alternatives for empty text.

pxlcoder committed:
[Cocoa] Inserting empty text using system text input interfaces does not delete the current selection
https://bugs.webkit.org/show_bug.cgi?id=263764
rdar://106570343

Reviewed by Wenson Hsieh.

Passing in an empty string to `-[NSResponder insertText:]` and
`-[UITextInput insertText:]` while there is a ranged selection should
perform backwards deletion, matching other system text engines.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::insertTextAsync):

`Editor::insertTextWithoutSendingTextEvent` early returns when empty text is
passed in. In order to avoid affecting other codepaths, implement the logic in
`WebPage::insertTextAsync`, which a common entry point for system text input
insertion across Cocoa platforms.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewTextInput.mm:
(insertText):
(TEST):

Canonical link: https://commits.webkit.org/269858@main
